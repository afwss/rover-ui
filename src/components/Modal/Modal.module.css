.modal {
  z-index: var(--rvr-zindex-modal-backdrop);
  position: fixed;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  background-color:rgba(0, 0, 0, 0.3);
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  justify-content: center;
  opacity: 0;
  transition: all 0.3s ease-in-out;
  overflow: auto;
  pointer-events: none;
}

.enterDone {
  opacity: 1;
  pointer-events: visible;
}

.exit {
  opacity: 0;
}

.modalContent {
  display: flex;
  flex-flow: column nowrap;
  z-index: var(--rvr-zindex-modal);
  background-color: var(--rvr-white);
  box-shadow: 0px 5px 10px rgba(0, 0, 0, 0.5);
  border-radius: 4px;
  transition: all 0.3s ease-in-out;
  overflow: auto;
  transform: translateY(-200px);
  margin: var(--rvr-space-bordered-lg);
  max-height: 100%;
  max-width: 100%;
}

.enterDone .modalContent {
  transform: translateY(0);
}

.exit .modalContent {
  transform: translateY(-200px);
}

.modalHeader {
  flex: 0 0 auto;
  border-radius: 4px 4px 0px 0px;
}

.modalHeader.level--primary {
  background: var(--rvr-gray-10);
  border-bottom: 1px solid var(--rvr-gray-20)
}

.modalHeader.level--warning {
  background: var(--rvr-yellow-lite-2);
  border-bottom: 1px solid var(--rvr-yellow);
}

.modalHeader.level--info {
  background: var(--rvr-blue-lite-2);
  border-bottom: 1px solid var(--rvr-blue);
}

.modalHeader.level--danger {
  background: var(--rvr-red-lite-2);
  border-bottom: 1px solid var(--rvr-red);
}


.modalFooter {
  flex: 0 0 auto;
  border-top: 1px solid var(--rvr-gray-20);
  background: var(--rvr-white);
}

.modalHeader, .modalFooter {
  padding: 12px;
}

.modalTitle {
  margin: 0;
}

.modalBody {
  flex: 1 1 auto;
  min-height: 0;
  overflow-y: auto;
  padding: 10px;
}

.bodyHasOpenModal {
  overflow: hidden;
}

.size--sm {
  width: 340px;
}	

.size--md {
  width: 578px;
}	

.size--lg {
  width: 815px;
}	

/*max-height: calc(var(--rvr-baseSize) * 40*/
@media (max-height: 320px) {
  .modalBody {
    flex: 0 0 auto;
  }

  .modalContent {
    flex: 0 0 auto;
    max-height: none;
  }

  .modal {
    justify-content: flex-start;
  }
}

