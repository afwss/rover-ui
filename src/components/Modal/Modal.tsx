import React, { CSSProperties, useCallback, useEffect } from 'react';
import { CSSTransition } from 'react-transition-group';
import classNames from 'classnames';
import { ClassValue } from 'classnames/types';

import ReactDOM from 'react-dom';
import styles from './Modal.module.css';

type Size = 'sm' | 'md' | 'lg';
type Level = 'primary' | 'warning' | 'info' | 'danger';

interface ModalProps {
  children?: React.ReactNode;
  className?: ClassValue;
  isOpen?: boolean;
  onClose?: () => void;
  size?: Size;
  style?: CSSProperties;
}

type ModalChildProps = Pick<ModalProps, 'children' | 'className'>;
type ModalHeaderProps = Pick<ModalProps, 'children' | 'className'> & {
  level?: Level;
};

type ModalType = React.FC<ModalProps> & {
  Header: React.FC<ModalHeaderProps>;
  Footer: React.FC<ModalChildProps>;
  Body: React.FC<ModalChildProps>;
};

const Modal: ModalType = ({
  children,
  className: passedClassName = '',
  isOpen = false,
  onClose = () => {},
  size = 'md',
  style: passedStyle = {},
  ...passedProps
}) => {
  const handleEscape = useCallback(
    (e: KeyboardEvent) => {
      if (isOpen && e.code === 'Escape' && onClose) {
        onClose();
      }
    },
    [isOpen, onClose]
  );

  useEffect(() => {
    if (isOpen) {
      window.addEventListener('keyup', handleEscape);
    }

    if (!isOpen) {
      window.removeEventListener('keyup', handleEscape);
    }

    return () => {
      if (isOpen) window.removeEventListener('keyup', handleEscape);
    };
  }, [handleEscape, isOpen]);

  useEffect(() => {
    if (isOpen) {
      document.body.className = classNames(
        document.body.className,
        styles.bodyHasOpenModal
      );
    } else {
      document.body.className = document.body.className
        .replace(styles.bodyHasOpenModal, '')
        .trim();
    }
  }, [isOpen]);

  const modalContentStyle = classNames(
    styles.content,
    {
      [styles[size]]: size,
    },
    passedClassName
  );

  return ReactDOM.createPortal(
    <CSSTransition
      in={isOpen}
      unmountOnExit
      timeout={{ enter: 0, exit: 300 }}
      classNames={{
        enterDone: styles.enterDone,
        exit: styles.exit,
      }}
    >
      <div
        role="presentation"
        {...passedProps}
        style={passedStyle}
        className={styles.Modal}
        onClick={onClose}
      >
        <div
          role="presentation"
          className={modalContentStyle}
          onClick={(e) => e.stopPropagation()}
        >
          {children}
        </div>
      </div>
    </CSSTransition>,
    document.body
  );
};

const Header: React.FC<ModalHeaderProps> = ({
  children,
  className = '',
  level = 'primary',
  ...props
}) => (
  <div
    {...props}
    className={classNames(
      styles.Header,
      { [styles[`level--${level}`]]: level },
      className
    )}
  >
    {children}
  </div>
);

const Body: React.FC<ModalChildProps> = ({
  children,
  className = '',
  ...props
}) => (
  <div {...props} className={classNames(styles.Body, className)}>
    {children}
  </div>
);

const Footer: React.FC<ModalChildProps> = ({
  children,
  className = '',
  ...props
}) => (
  <div {...props} className={classNames(styles.Footer, className)}>
    {children}
  </div>
);

Modal.Header = Header;
Modal.Body = Body;
Modal.Footer = Footer;

export default Modal;
