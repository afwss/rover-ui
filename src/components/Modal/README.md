# \<Modal\>

**A configurable Modal component that automatically triggers onClose when escape key is pressed or clicked outside**

This component can be used as a dialog box that slides in from the top.

Below is an example structure of the `<Modal>`

```
<Modal>
  <Modal.Header /> (optional)
  <Modal.Body> --- main content of the modal goes here --- </Modal.Body>
  <Modal.Footer /> (optional)
</Modal>
```

This will render a modal with a header at the top, a footer at the bottom, and a body block that scrolls if the interior content extends outside the height of the parent.
