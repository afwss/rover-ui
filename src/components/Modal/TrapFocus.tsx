import React, { useCallback, useEffect } from 'react';

const getFocusableElements = (element: HTMLElement) => {
  return element?.querySelectorAll(
    'a, button, textarea, input, input[type="radio"], input[type="checkbox"], select'
  );
};

const getFirstFocusableElement = (element: HTMLElement) => {
  const focusableElements = getFocusableElements(element);
  return focusableElements && focusableElements[0];
};

const getLastFocusableElement = (element: HTMLElement) => {
  const focusableElements = getFocusableElements(element);
  return focusableElements && focusableElements[focusableElements.length - 1];
};

const TrapFocus = ({ elementRef, isOpen = true }) => {
  console.log('Element coming in ', elementRef);
  const handleTabKey = useCallback(
    (e: KeyboardEvent) => {
      const element = elementRef?.current;
      console.log('Tabbing ');
      const focusableElements = getFocusableElements(element);
      if (focusableElements) {
        const firstFocusableElement = getFirstFocusableElement(element);
        const lastFocusableElement = getLastFocusableElement(element);

        if (!e.shiftKey && document.activeElement === lastFocusableElement) {
          console.log('Tabbing in not shiftkey ');
          (firstFocusableElement as HTMLElement)?.focus();
          return e.preventDefault();
        }

        if (e.shiftKey && document.activeElement === firstFocusableElement) {
          console.log('Tabbing with shiftkey');
          (lastFocusableElement as HTMLElement)?.focus();
          return e.preventDefault();
        }
      }

      return null;
    },
    [elementRef]
  );

  useEffect(() => {
    console.log('Trap focus use effect');
    const keyListenerMap = new Map([['Tab', handleTabKey]]);
    const keyListener = (e: KeyboardEvent) => {
      console.log('wiring up tab');
      const listener = keyListenerMap.get(e.code);
      return listener && listener(e);
    };

    if (isOpen) {
      window.addEventListener('keydown', keyListener);
    }

    if (!isOpen) {
      window.removeEventListener('keydown', keyListener);
    }

    return () => {
      if (isOpen) window.removeEventListener('keydown', keyListener);
    };
  }, [handleTabKey, isOpen]);

  return null;
};

const AutoFocusElement = ({ element }) => {
  useEffect(() => {
    const firstElement = getFirstFocusableElement(element);
    (firstElement as HTMLElement)?.focus();
  }, [element]);
};

export { TrapFocus, AutoFocusElement };
