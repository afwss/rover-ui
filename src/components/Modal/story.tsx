import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { Wrap } from '../../stories/storybook-helpers';

import Modal from '.';
import Readme from './README.md';

import Button from '../Button';
import Typography from '../Typography';
import Input from '../Input';

function SmallModalExample() {
  const [isOpen, setIsOpen] = useState(false);

  const openModal = () => {
    setIsOpen(true);
  };

  const closeModal = () => {
    setIsOpen(false);
  };

  return (
    <>
      <div>
        <div>Small modal</div>
        <Button modifiers={['primary']} onClick={openModal}>
          Show Modal
        </Button>
      </div>
      <Modal isOpen={isOpen} onClose={closeModal} size="sm">
        <Modal.Header>
          <h4 className="">Modal Title</h4>
        </Modal.Header>
        <Modal.Body>
          <p>This is the modal body</p>
          <p>This is the modal body</p>
          <p>This is the modal body</p>
          <p>This is the modal body</p>
          <p>This is the modal body</p>
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={closeModal} className="">
            Close
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}

const MediumModalExample = () => {
  const [isOpen, setIsOpen] = useState(false);

  const openModal = () => {
    setIsOpen(true);
  };

  const closeModal = () => {
    setIsOpen(false);
  };

  return (
    <>
      <div>
        <div>Medium modal</div>
        <Button modifiers={['primary']} onClick={openModal}>
          Show Modal
        </Button>
      </div>
      <Modal isOpen={isOpen} onClose={closeModal} size="md">
        <Modal.Header level="warning">
          <h4 className="">Modal Title</h4>
        </Modal.Header>
        <Modal.Body>
          <p>This is the modal body</p>
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={closeModal} className="">
            Close
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
};

const LargeModalExample = () => {
  const [isOpen, setIsOpen] = useState(false);
  const initialInputValues = {
    firstName: '',
    lastName: '',
    email: '',
    password: '',
  };
  const [formInputValues, setFormInputValues] = useState(initialInputValues);

  const openModal = () => {
    setIsOpen(true);
  };

  const closeModal = () => {
    setIsOpen(false);
    setFormInputValues({
      ...initialInputValues,
    });
  };

  const handleInputChange = (event) => {
    event.persist();

    setFormInputValues({
      ...formInputValues,
      [event.target.name]: event.target.value,
    });
  };

  return (
    <>
      <div>
        <div>large modal example</div>
        <Button modifiers={['primary']} onClick={openModal}>
          Show Modal
        </Button>
      </div>
      <Modal isOpen={isOpen} onClose={closeModal} size="lg">
        <Modal.Header level="info">
          <Typography tag="h4" size="xl" color="black">
            Sign up Form
          </Typography>
        </Modal.Header>
        <Modal.Body>
          <div className="mb-6">
            <div className="flex flex-col mb-4">
              <Typography size="lg">First Name</Typography>
              <Input
                autoComplete="off"
                className="border py-2 px-3 text-grey-darkest"
                type="text"
                name="firstName"
                placeholder="First Name"
                onChange={handleInputChange}
                value={formInputValues.firstName}
              />
            </div>
            <div className="flex flex-col mb-4">
              <Typography size="lg">Last Name</Typography>
              <Input
                autoComplete="off"
                className="border py-2 px-3 text-grey-darkest"
                type="text"
                name="lastName"
                placeholder="Last Name"
                onChange={handleInputChange}
                value={formInputValues.lastName}
              />
            </div>
            <div className="flex flex-col mb-4">
              <Typography size="lg">Email</Typography>
              <Input
                autoComplete="off"
                className="border py-2 px-3 text-grey-darkest"
                type="text"
                name="email"
                placeholder="Email"
                onChange={handleInputChange}
                value={formInputValues.email}
              />
            </div>
            <div className="flex flex-col mb-6">
              <Typography size="lg">Password</Typography>
              <Input
                autoComplete="off"
                className="border py-2 px-3 text-grey-darkest"
                type="password"
                name="password"
                onChange={handleInputChange}
                value={formInputValues.password}
              />
            </div>
          </div>
        </Modal.Body>
        <Modal.Footer>
          <div>
            <Button className="t-mr" level="tertiary" onClick={closeModal}>
              Cancel
            </Button>
            <Button level="success" onClick={closeModal}>
              Sign Up!
            </Button>
          </div>
        </Modal.Footer>
      </Modal>
    </>
  );
};

storiesOf('Galaxies/Modal', module)
  .addParameters({
    readme: {
      sidebar: Readme,
    },
  })
  .add(
    'Overview',
    () => (
      <Modal isOpen onClose={() => {}}>
        <Modal.Header>
          <h4 className="">Modal Title</h4>
        </Modal.Header>
        <Modal.Body>
          <p>This is the modal body</p>
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => {}} className="">
            Close
          </Button>
        </Modal.Footer>
      </Modal>
    ),
    {
      info: {
        inline: true,
        source: true,
      },
    }
  )
  .add(
    'Examples',
    () => (
      <>
        <Wrap>
          <SmallModalExample />
        </Wrap>
        <Wrap>
          <MediumModalExample />
        </Wrap>
        <Wrap>
          <LargeModalExample />
        </Wrap>
      </>
    ),
    {
      info: {
        inline: false,
        source: false,
      },
    }
  );
