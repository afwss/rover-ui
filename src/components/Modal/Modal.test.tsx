import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Modal from './Modal';

describe('Modal', () => {
  it('renders', () => {
    render(<Modal isOpen data-testid="Modal-Test" />);
    expect(screen.getByTestId('Modal-Test')).toBeInTheDocument();
  });

  it("does not render when 'isOpen' prop is false", () => {
    render(<Modal isOpen={false} data-testid="Modal-Test" />);
    expect(screen.queryByTestId('Modal-Test')).not.toBeInTheDocument();
  });

  describe('Modal CSS Classes', () => {
    test.each`
      size         | level        | sizeClass | levelClass
      ${'sm'}      | ${'primary'} | ${'sm'}   | ${'level--primary'}
      ${'md'}      | ${'warning'} | ${'md'}   | ${'level--warning'}
      ${'lg'}      | ${undefined} | ${'lg'}   | ${'level--primary'}
      ${undefined} | ${undefined} | ${'md'}   | ${'level--primary'}
      ${'lg'}      | ${'info'}    | ${'lg'}   | ${'level--info'}
      ${undefined} | ${'danger'}  | ${'md'}   | ${'level--danger'}
    `(
      'when size prop = $size and level prop = $level, the modal should have css classes $sizeClass and $levelClass',
      ({ size, level, sizeClass, levelClass }) => {
        render(
          <Modal isOpen size={size}>
            <Modal.Header data-testid="Modal-Header" level={level}>
              <h4>Test Header</h4>
            </Modal.Header>
          </Modal>
        );
        const modalContentDiv = screen.getAllByRole('presentation')[1];
        expect(modalContentDiv).toHaveClass(sizeClass);
        expect(screen.getByTestId('Modal-Header')).toHaveClass(levelClass);
      }
    );
  });

  describe('onClose callback', () => {
    it('calls onClose callback when the escape key is pressed', async () => {
      const onClose = jest.fn();
      render(<Modal isOpen onClose={onClose} data-testid="Modal-Test" />);
      const modal = screen.getByTestId('Modal-Test');
      expect(modal).toBeInTheDocument();
      fireEvent.keyUp(modal, {
        key: 'Escape',
        code: 'Escape',
      });
      expect(onClose).toHaveBeenCalledTimes(1);
    });
  });
});
