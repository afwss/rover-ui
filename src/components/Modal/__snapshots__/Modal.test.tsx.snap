// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<body
  class="bodyHasOpenModal"
>
  <div />
  <div
    class="Modal"
    role="presentation"
  >
    <div
      class="content md"
      role="dialog"
    >
      <div
        class="Header level--primary"
      >
        <p>
          Modal Header
        </p>
      </div>
      <div
        class="Body"
      >
        <p>
          Modal Body
        </p>
      </div>
      <div
        class="Footer"
      >
        <p>
          Modal Footer
        </p>
      </div>
    </div>
  </div>
</body>
`;
